walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,140 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.util.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request == null) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    if (events == null || events.size() <= 0) {
+      if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return new ArrayList<>();
+      } else {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+    }
+
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> attendeesWithOptional = new ArrayList<>();
+    attendeesWithOptional.addAll(request.getAttendees());
+    attendeesWithOptional.addAll(request.getOptionalAttendees());
+
+    long duration = request.getDuration();
+
+    List<TimeRange> occupiedTimeRange = new ArrayList<>();
+    List<TimeRange> occupiedTimeRangeWithOptional = new ArrayList<>();
+
+    for (Event event: events) {
+      if (hasAttendeesAttending(event.getAttendees(), attendees)) {
+        occupiedTimeRange.add(event.getWhen());
+      }
+
+      if (hasAttendeesAttending(event.getAttendees(), attendeesWithOptional)) {
+        occupiedTimeRangeWithOptional.add(event.getWhen());
+      }
+    }
+
+    Collection<TimeRange> availableMeetings = getAvailableTimeRange(occupiedTimeRangeWithOptional, duration);
+
+    if (!availableMeetings.isEmpty()) {
+      return availableMeetings;
+    } else {
+      // Has mandatory attendees
+      if (!attendees.isEmpty()) {
```

Comment:
should you move that check earlier? No point computing meetings if there are no attendes